kernel/signal.c

```diff
@@ -0,0 +1,552 @@
+}
+#include <syscalls/k_sig_timedwait_mrsh.c>
+#endif
+
+int z_impl_k_sig_mask(int how, const struct k_sig_set *set, struct k_sig_set *oset)
+{
+	int ret = 0;
+	k_tid_t tid = k_current_get();
+
+	if (!(how == K_SIG_BLOCK || how == K_SIG_SETMASK || how == K_SIG_UNBLOCK)) {
+		return -EINVAL;
+	}
+
+	if (oset != NULL) {
+		*oset = tid->base.sig_mask;
+	}
+
+	if (set == NULL) {
+		return 0;
+	}
+
+	switch (how) {
+	case K_SIG_BLOCK:
+		k_sig_orset(&tid->base.sig_mask, set);
+		break;
+	case K_SIG_SETMASK:
+		tid->base.sig_mask = *set;
+		break;
+	case K_SIG_UNBLOCK:
+		k_sig_unset(&tid->base.sig_mask, set);
+		break;
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_sig_mask(int how, const struct k_sig_set *set, struct k_sig_set *oset)
+{
+	if (set != NULL) {
+		K_OOPS(K_SYSCALL_MEMORY_READ(set, sizeof(*set)));
+	}
+
+	if (oset != NULL) {
+		K_OOPS(K_SYSCALL_MEMORY_WRITE(oset, sizeof(*oset)));
+	}
+
+	return z_impl_k_sig_mask(how, set, oset);
+}
+#include <syscalls/k_sig_mask_mrsh.c>
+#endif
+
+int z_impl_k_sig_pending(struct k_sig_set *set)
+{
+	K_SPINLOCK(k_sig_lock) {
+		k_sig_pending_unlocked(set);
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_sig_pending(struct k_sig_set *set)
+{
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(set, sizeof(*set)));
+	return z_impl_k_sig_pending(set);
+}
+#include <syscalls/k_sig_pending_mrsh.c>
+#endif
+
+int k_sig_addset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] |= BIT(bit);
+
+	return 0;
+}
+
+int k_sig_delset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] &= ~BIT(bit);
+
+	return 0;
+}
+
+int k_sig_ismember(const struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	return (set->sig[i] & BIT(bit)) != 0;
+}
+
+int k_sig_fillset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = -1;
+	}
+
+	return 0;
+}
+
+int k_sig_emptyset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = 0;
+	}
+
+	return 0;
+}
+
+static void k_sig_orset(struct k_sig_set *result, const struct k_sig_set *set2)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] |= set2->sig[i];
+	}
+}
+
+static void k_sig_unset(struct k_sig_set *result, const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] &= ~set->sig[i];
+	}
+}
+
+static bool k_sig_setisempty(const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(set->sig, i) {
+		if (POPCOUNT(set->sig[i]) > 0) {
```

Comment:
Can't you just check `if (set->sig[i] > 0)`

Reply:
Yes - fixed!